include/functional/detail/meta.hpp

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2024 Gašper Ažman, Bronek Kozicki
+//
+// Distributed under the ISC License. See accompanying file LICENSE.md
+// or copy at https://opensource.org/licenses/ISC
+
+#ifndef INCLUDE_FUNCTIONAL_DETAIL_META
+#define INCLUDE_FUNCTIONAL_DETAIL_META
+
+#include <algorithm>
+#include <array>
+#include <concepts>
+#include <cstdint>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+namespace fn::detail {
+
+// TODO Remove `select_nth` when our compilers start supporting pack indexing https://wg21.link/P2662
+template <std::size_t, typename...> struct select_nth;
+template <std::size_t N> struct select_nth<N>; // Intentionally incomplete type
+
+template <std::size_t N, typename T1, typename... Ts> struct select_nth<N, T1, Ts...> {
+  static_assert(N < (1 + sizeof...(Ts)));
+  using type = std::tuple_element_t<N, std::tuple<T1, Ts...>>;
+};
+
+template <std::size_t N, typename... Ts> using select_nth_t = select_nth<N, Ts...>::type;
+
+// NOTE Reverse to above, i.e. mapping of a type to position in a list of types.
+template <typename T> struct _indexed_type {
+  std::size_t index;
+};
+template <typename... Ts> struct _indexed_type_list : _indexed_type<Ts>... {
+  constexpr _indexed_type_list(std::size_t i = 0) : _indexed_type<Ts>{i++}... {}
+};
+template <typename T, typename... Ts>
+  requires(... || std::same_as<Ts, T>)
+constexpr inline std::size_t type_index = static_cast<_indexed_type<T> const &>(_indexed_type_list<Ts...>()).index;
+
+template <typename T, typename... Ts> constexpr inline bool type_one_of = (... || std::same_as<Ts, T>);
+
+template <auto TU_name, auto Input> struct _normalized_name final {
+  static constexpr std::size_t bound = 4096;
```

Comment:
this has to be rewritten to not need 4k of memory per type...

Reply:
done, not sure if you will like it (but will probably make you laugh)